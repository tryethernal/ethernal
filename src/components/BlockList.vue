<template>
    <v-data-table
        dense
        :loading="loading"
        :items="blocks"
        :sort-by="currentOptions.sortBy[0]"
        :must-sort="true"
        :sort-desc="true"
        :server-items-length="blockCount"
        :hide-default-footer="dense"
        :hide-default-header="dense"
        :footer-props="{
            itemsPerPageOptions: [10, 25, 100]
        }"
        :headers="headers"
        @update:options="getBlocks">
        <template v-slot:no-data>
            No blocks found
        </template>
        <template v-slot:item.number="{ item }">
            <router-link :to="'/block/' + item.number">{{item.number}}</router-link>
        </template>
        <template v-slot:item.timestamp="{ item }">
            <div class="my-2 text-left">
                {{ moment(item.timestamp) | moment('MM/DD h:mm:ss A') }}<br>
                <small>{{ moment(item.timestamp).fromNow() }}</small>
            </div>
        </template>
        <template v-slot:item.gasUsed="{ item }">
            {{ item.gasUsed.toLocaleString()  }}
        </template>
        <template v-slot:item.transactionNumber="{ item }">
            {{ item.transactionsCount  }} {{ item.transactionsCount != 1 ? 'transactions' : 'transaction' }}
        </template>
    </v-data-table>
</template>

<script>
const moment = require('moment');
export default {
    name: 'BlockList',
    props: ['dense'],
    data: () => ({
        blocks: [],
        blockCount: 0,
        headers: [],
        loading: true,
        currentOptions: { page: 1, itemsPerPage: 10, sortBy: ['number'], sortDesc: [true] }
    }),
    mounted: function() {
        this.pusher.onNewBlock(() => this.getBlocks(this.currentOptions), this);
        if (this.dense) {
            this.headers = [
                {
                    text: 'Block',
                    value: 'number'
                },
                {
                    text: 'Mined On',
                    value: 'timestamp'
                },
                {
                    text: 'Transaction Count',
                    value: 'transactionNumber',
                    sortable: false
                }
            ]
        }
        else {
            this.headers = [
                {
                    text: 'Block',
                    value: 'number'
                },
                {
                    text: 'Mined On',
                    value: 'timestamp'
                },
                {
                    text: 'Gas Used',
                    value: 'gasUsed',
                    sortable: false
                },
                {
                    text: 'Transaction Count',
                    value: 'transactionNumber',
                    sortable: false
                }
            ]
        }
    },
    methods: {
        moment: moment,
        getBlocks: function(newOptions) {
            this.loading = true;

            if (newOptions)
                this.currentOptions = newOptions;

            const options = {
                page: this.currentOptions.page,
                itemsPerPage: this.currentOptions.itemsPerPage,
                order: this.currentOptions.sortDesc[0] === false ? 'asc' : 'desc'
            };
            this.server.getBlocks(options)
                .then(({ data }) => {
                    this.blocks = data.items;
                    this.blockCount = data.total;
                })
                .catch(console.log)
                .finally(() => this.loading = false);
        }
    }
}
</script>
