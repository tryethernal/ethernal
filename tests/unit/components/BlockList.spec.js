import MockHelper from '../MockHelper';

import BlockList from '@/components/BlockList.vue';

describe('BlockList.vue', () => {
    let helper, blocks;

    beforeEach(() => {
        helper = new MockHelper();
        jest.spyOn(Date, 'now').mockImplementation(() => new Date('2022-08-07T12:33:37.000Z'));
    });

    it('Should display pending blocks', async () => {
        const blocks = [
            {
                number: '2',
                gasUsed: '200000',
                timestamp: '2022-05-06T17:11:26.000Z',
                hash: '0x99c6edb3bb1124680a97661c1f5794d60617abb57bd1e611d81fc5b941f36d30',
                state: 'syncing'
            },
            {
                number: '1',
                gasUsed: '100000',
                timestamp: '2022-05-06T17:11:26.000Z',
                hash: '0x98c6edb3bb1124680a97661c1f5794d60617abb57bd1e611d81fc5b941f36d30',
                transactions: [
                    {
                        hash: '0x060034486a819816df57d01eefccbe161d7019f9f3c235e18af07468fb194ef0',
                        timestamp: '2022-05-06T17:11:26.000Z',
                        from: '0x0',
                        to: 'Ox1',
                        blockNumber: 1,
                        value: '0'
                    }
                ]
            }
        ];

        jest.spyOn(helper.mocks.server, 'getBlocks')
            .mockResolvedValue({ data: { items: blocks, blockCount: 2 }});

        const wrapper = helper.mountFn(BlockList);
        await new Promise(process.nextTick);

        expect(wrapper.vm.blocks).toEqual(blocks);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show the blocks list', async () => {
        const blocks = [
            {
                number: '2',
                gasUsed: '200000',
                timestamp: '2022-05-06T17:11:26.000Z',
                hash: '0x99c6edb3bb1124680a97661c1f5794d60617abb57bd1e611d81fc5b941f36d30'                
            },
            {
                number: '1',
                gasUsed: '100000',
                timestamp: '2022-05-06T17:11:26.000Z',
                hash: '0x98c6edb3bb1124680a97661c1f5794d60617abb57bd1e611d81fc5b941f36d30',
                transactions: [
                    {
                        hash: '0x060034486a819816df57d01eefccbe161d7019f9f3c235e18af07468fb194ef0',
                        timestamp: '2022-05-06T17:11:26.000Z',
                        from: '0x0',
                        to: 'Ox1',
                        blockNumber: 1,
                        value: '0'
                    }
                ]
            }
        ];

        jest.spyOn(helper.mocks.server, 'getBlocks')
            .mockResolvedValue({ data: { items: blocks, blockCount: 2 }});

        const wrapper = helper.mountFn(BlockList);
        await new Promise(process.nextTick);

        expect(wrapper.vm.blocks).toEqual(blocks);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should show the loading message when empty blocks list', async () => {
        jest.spyOn(helper.mocks.server, 'getBlocks')
            .mockResolvedValue({ data: { items: [], blockCount: 0 }});

        const wrapper = helper.mountFn(BlockList);
        await new Promise(process.nextTick);

        expect(wrapper.html()).toMatchSnapshot();
    });
});
