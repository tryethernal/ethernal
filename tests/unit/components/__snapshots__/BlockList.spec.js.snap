// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockList.vue Should display pending blocks 1`] = `
<div class="v-data-table v-data-table--dense v-data-table--has-bottom theme--light">
  <div class="v-data-table__wrapper">
    <table>
      <colgroup>
        <col class="">
        <col class="">
        <col class="">
        <col class="">
      </colgroup>
      <thead class="v-data-table-header">
        <tr>
          <th role="columnheader" scope="col" aria-label="Block: Sorted descending. Activate to sort ascending." aria-sort="descending" class="text-start sortable active desc"><span>Block</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          <th role="columnheader" scope="col" aria-label="Mined On: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Mined On</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          <th role="columnheader" scope="col" aria-label="Gas Used" class="text-start"><span>Gas Used</span></th>
          <th role="columnheader" scope="col" aria-label="Transaction Count" class="text-start"><span>Transaction Count</span></th>
        </tr>
      </thead>
      <tbody>
        <tr class="isSyncing">
          <td class="text-start">
            <div role="progressbar" aria-valuemin="0" aria-valuemax="100" class="v-progress-circular mr-2 v-progress-circular--visible v-progress-circular--indeterminate primary--text" style="height: 16px; width: 16px;" aria-haspopup="true" aria-expanded="false"><svg xmlns="http://www.w3.org/2000/svg" viewBox="22.857142857142858 22.857142857142858 45.714285714285715 45.714285714285715" style="transform: rotate(0deg);">
                <circle fill="transparent" cx="45.714285714285715" cy="45.714285714285715" r="20" stroke-width="5.714285714285714" stroke-dasharray="125.664" stroke-dashoffset="125.66370614359172px" class="v-progress-circular__overlay"></circle>
              </svg>
              <div class="v-progress-circular__info"></div>
            </div><span class="v-tooltip v-tooltip--top"><!----></span> <a href="#/block/2" class="">2</a>
          </td>
          <td class="text-start">
            <div class="my-2 text-left">
              05/06 7:11:26 PM<br> <small>3 months ago</small></div>
          </td>
          <td class="text-start">
            200000
          </td>
          <td class="text-start">
            transactions
          </td>
        </tr>
        <tr class="">
          <td class="text-start"><span class="v-tooltip v-tooltip--top"><!----></span> <a href="#/block/1" class="">1</a></td>
          <td class="text-start">
            <div class="my-2 text-left">
              05/06 7:11:26 PM<br> <small>3 months ago</small></div>
          </td>
          <td class="text-start">
            100000
          </td>
          <td class="text-start">
            transactions
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div class="v-data-footer">
    <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-11" class="v-input__slot">
            <div class="v-select__slot">
              <div class="v-select__selections">
                <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Rows per page:" id="input-11" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
              </div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
              </div><input type="hidden" value="10">
            </div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="v-data-footer__pagination">1-2 of 2</div>
    <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
    <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
  </div>
</div>
`;

exports[`BlockList.vue Should show the blocks list 1`] = `
<div class="v-data-table v-data-table--dense v-data-table--has-bottom theme--light">
  <div class="v-data-table__wrapper">
    <table>
      <colgroup>
        <col class="">
        <col class="">
        <col class="">
        <col class="">
      </colgroup>
      <thead class="v-data-table-header">
        <tr>
          <th role="columnheader" scope="col" aria-label="Block: Sorted descending. Activate to sort ascending." aria-sort="descending" class="text-start sortable active desc"><span>Block</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          <th role="columnheader" scope="col" aria-label="Mined On: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Mined On</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          <th role="columnheader" scope="col" aria-label="Gas Used" class="text-start"><span>Gas Used</span></th>
          <th role="columnheader" scope="col" aria-label="Transaction Count" class="text-start"><span>Transaction Count</span></th>
        </tr>
      </thead>
      <tbody>
        <tr class="">
          <td class="text-start"><span class="v-tooltip v-tooltip--top"><!----></span> <a href="#/block/2" class="">2</a></td>
          <td class="text-start">
            <div class="my-2 text-left">
              05/06 7:11:26 PM<br> <small>3 months ago</small></div>
          </td>
          <td class="text-start">
            200000
          </td>
          <td class="text-start">
            transactions
          </td>
        </tr>
        <tr class="">
          <td class="text-start"><span class="v-tooltip v-tooltip--top"><!----></span> <a href="#/block/1" class="">1</a></td>
          <td class="text-start">
            <div class="my-2 text-left">
              05/06 7:11:26 PM<br> <small>3 months ago</small></div>
          </td>
          <td class="text-start">
            100000
          </td>
          <td class="text-start">
            transactions
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div class="v-data-footer">
    <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-36" class="v-input__slot">
            <div class="v-select__slot">
              <div class="v-select__selections">
                <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Rows per page:" id="input-36" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
              </div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
              </div><input type="hidden" value="10">
            </div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="v-data-footer__pagination">1-2 of 2</div>
    <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
    <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
  </div>
</div>
`;

exports[`BlockList.vue should show the loading message when empty blocks list 1`] = `
<div class="v-data-table v-data-table--dense v-data-table--has-bottom theme--light">
  <div class="v-data-table__wrapper">
    <table>
      <colgroup>
        <col class="">
        <col class="">
        <col class="">
        <col class="">
      </colgroup>
      <thead class="v-data-table-header">
        <tr>
          <th role="columnheader" scope="col" aria-label="Block: Sorted descending. Activate to sort ascending." aria-sort="descending" class="text-start sortable active desc"><span>Block</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          <th role="columnheader" scope="col" aria-label="Mined On: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Mined On</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          <th role="columnheader" scope="col" aria-label="Gas Used" class="text-start"><span>Gas Used</span></th>
          <th role="columnheader" scope="col" aria-label="Transaction Count" class="text-start"><span>Transaction Count</span></th>
        </tr>
      </thead>
      <tbody>
        <tr class="v-data-table__empty-wrapper">
          <td colspan="4">
            No blocks found
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div class="v-data-footer">
    <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-60" class="v-input__slot">
            <div class="v-select__slot">
              <div class="v-select__selections">
                <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Rows per page:" id="input-60" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
              </div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
              </div><input type="hidden" value="10">
            </div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="v-data-footer__pagination">–</div>
    <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
    <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
  </div>
</div>
`;
